crates/oxc_linter/src/rules/typescript/ban_types.rs

```diff
@@ -0,0 +1,162 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum BanTypesDiagnostic {
+    #[error("eslint@typescript-eslint/ban-types: Do not use `{0}` as a type. Use `{1}` instead.")]
+    Type(String, String, #[label] Span),
+    #[error("eslint@typescript-eslint/ban-types: Prefer explicitly define the object shape. This type means \"any non-nullish value\", which is slightly better than 'unknown', but it's still a broad type.")]
+    TypeLiteral(#[label] Span),
+    #[error("eslint@typescript-eslint/ban-types: Don't use `Function` as a type. The `Function` type accepts any function-like value.
+    It provides no type safety when calling the function, which can be a common source of bugs.
+    It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
+    If you are expecting the function to accept certain arguments, you should explicitly define the function shape.")]
+    Function(#[label] Span),
+    #[error(r#"eslint@typescript-eslint/ban-types: 'The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.',
+    - If you want a type meaning "any object", you probably want `object` instead.
+    - If you want a type meaning "any value", you probably want `unknown` instead.
+    - If you really want a type meaning "any non-nullish value", you probably want `NonNullable<unknown>` instead."#)]
+    Object(#[label] Span),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct BanTypes;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// This rule bans specific types and can suggest alternatives. Note that it does not ban the corresponding runtime objects from being used.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Some built-in types have aliases, while some types are considered dangerous or harmful. It's often a good idea to ban certain types to help with consistency and safety.
+    ///
+    /// ### Example
+    /// ```typescript
+    /// let foo: String = 'foo';
+    ///
+    /// let bar: Boolean = true;
+    /// ```
+    BanTypes,
+    correctness
+);
+
+impl Rule for BanTypes {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::TSTypeReference(typ) => {
+                let name = match &typ.type_name {
+                    oxc_ast::ast::TSTypeName::IdentifierReference(v) => &v.name,
+                    oxc_ast::ast::TSTypeName::QualifiedName(_) => return,
+                };
+
+                match name.as_str() {
+                    "String" | "Boolean" | "Number" | "Symbol" | "BigInt" => {
+                        ctx.diagnostic(BanTypesDiagnostic::Type(
```

Comment:
Consider using `&'a Atom` here to reduce memory allocations

Reply:
instead of `name.to_string` or the match?

Reply:
The former, I misclicked the line

Reply:
:+1: thanks. mind checking now?